data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -0,0 +1,163 @@
+
+    private final List<ScanOptions> scanOptionsBuilderList;
+
+    private final S3ObjectHandler s3ObjectHandler;
+
+    private final BucketOwnerProvider bucketOwnerProvider;
+
+    public ScanObjectWorker(final S3Client s3Client,
+                            final List<ScanOptions> scanOptionsBuilderList,
+                            final S3ObjectHandler s3ObjectHandler,
+                            final BucketOwnerProvider bucketOwnerProvider){
+        this.s3Client = s3Client;
+        this.scanOptionsBuilderList = scanOptionsBuilderList;
+        this.s3ObjectHandler= s3ObjectHandler;
+        this.bucketOwnerProvider = bucketOwnerProvider;
+    }
+
+    /**
+     * It will decide the s3 object parse <code>S3ObjectWorker</code> or <code>S3SelectWorker</code>
+     * based on s3 select configuration provided.
+     */
+    @Override
+    public void run() {
+        scanOptionsBuilderList.forEach(this::parseS3ScanObjects);
+    }
+
+    /**
+     * Method will parse the s3 object and write to {@link Buffer}
+     */
+    void parseS3ScanObjects(final ScanOptions scanOptions) {
+        final List<String> scanObjects = new ArrayList<>();
+        final List<String> excludeItems = new ArrayList<>();
+        final S3ScanKeyPathOption s3ScanKeyPathOption = scanOptions.getS3ScanKeyPathOption();
+        final ListObjectsV2Request.Builder listObjectsV2Request = ListObjectsV2Request.builder()
+                .bucket(scanOptions.getBucket());
+        bucketOwnerProvider.getBucketOwner(scanOptions.getBucket())
+                .ifPresent(listObjectsV2Request::expectedBucketOwner);
+
+        if(Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3ScanExcludeOptions()))
+            excludeItems.addAll(s3ScanKeyPathOption.getS3ScanExcludeOptions());
+
+        if(Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3scanIncludeOptions()))
+            s3ScanKeyPathOption.getS3scanIncludeOptions().stream().forEach(includePath -> {
+                listObjectsV2Request.prefix(includePath);
+                scanObjects.addAll(listS3Objects(excludeItems, listObjectsV2Request));
+            });
+        else
+            scanObjects.addAll(listS3Objects(excludeItems, listObjectsV2Request));
+
+        if(scanObjects.isEmpty())
+            LOG.info("s3 objects are not found in configured scan pipeline.");
+
+        scanObjects.stream().forEach(key ->
+                processS3ObjectKeys(S3ObjectReference.bucketAndKey(scanOptions.getBucket(),
+                        key).build(),s3ObjectHandler, scanOptions));
+    }
+
+    private List<String> listS3Objects(List<String> excludeKeyPaths, ListObjectsV2Request.Builder listObjectsV2Request) {
+        return s3Client.listObjectsV2(listObjectsV2Request.build()).contents().stream().map(S3Object::key)
+                .filter(s3ObjKey-> s3ObjKey.lastIndexOf(".")!=-1)
```

Comment:
Why is this check done for all objects ? Will all the S3ObjectKey contain `"."` ?

Reply:
Good catch. Not all keys will have a `.`.

Also, the filter on the next two lines will match on `test-jpg.txt` when excluding `jpg`, even though that is not what we want.

I believe you can replace lines 101-103 with:

```
.filter(s3ObjKey -> !excludeKeyPaths.stream()
                        .anyMatch(excludeItem -> s3ObjKey.endsWith("." + excludeItem)))
```

Reply:
Thanks for your review @dlvenable @dinujoh, we have modified the code as you suggested above but "." was removed so that user can configure either **csv** or **.csv** in **exclude_suffix:** items. 

@dinujoh we did "." filter because of directory paths are also coming in **s3Client.listObjectsV2()** output.so as david mentioned "." might not come for some files right so i was handled in different way. Please review it.